### PR TITLE
half2 type

### PR DESCRIPTION
**Pull request by ap-hynninen**
Implements half2 type that is needed for efficient half precision computations. MSHADOW_TYPE_SWITCH_WITH_HALF2 converts float16 input into half2

### CODE REVIEW

File: mshadow/half2.h [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    half_t2[0] = a;
+    half_t2[1] = b;
+  }
+#endif
+
+  MSHADOW_HALF2_INLINE explicit half2_t(int a) {
+#if MSHADOW_CUDA_HALF2
+    half2_ = __half2half2(__int2half_rz(a));
+#else
+    half_t2[0] = (half_t)a;
```
Comment (piiswrong) [inline mshadow/half2.h:54]: Are there endian issues?

Comment (ap-hynninen) [reply mshadow/half2.h:54]: Yes, I think there is if we mix half2 computation on CPU and GPU and they have different endians.

**Comment by ap-hynninen** [comment]
What more do we need here to get this in?

**Comment by piiswrong** [comment]
please resolve conflict

**Comment by piiswrong** [comment]
Btw can we change pack to lanes? It's more consistent with dlpack

**Pull request merged**